### PR TITLE
Add cross-build support for arm64 on x64 hosts

### PR DESCRIPTION
**Pull request by TomohideMorimoto**
# Description
Added a toolchain file for cross-build. This file is a sample for aarch64 when using the Yocto SDK. Added usage to meta-flutter / README.md.

## Related issues
#57

### CODE REVIEW

**Comment by TomohideMorimoto** [comment]
@HidenoriMatsubayashi Can I request you to review?

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,9 +1,11 @@
 cmake_minimum_required(VERSION 3.10)
 project("flutter_linuxes" LANGUAGES CXX C)
 
-# Rules of compiler and cpp version.
-set(CMAKE_C_COMPILER clang)
-set(CMAKE_CXX_COMPILER clang++)
+# Rules of cpp version.
```
Comment (HidenoriMatsubayashi) [inline CMakeLists.txt:4]: The comment is not appropriate.

File: cmake/cross-toolchain.cmake [hunk 2]
```diff
@@ -0,0 +1,32 @@
+cmake_minimum_required(VERSION 3.10)
```
Comment (HidenoriMatsubayashi) [inline cmake/cross-toolchain.cmake:1]: Please change the filename to cross-toolchain-aarch64-yocto-template.cmake

File: cmake/cross-toolchain.cmake [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+# Specify the sysroot
+set(sdk_target_sysroot "/opt/poky/3.1.7/sysroots/aarch64-poky-linux")
+set(CMAKE_SYSROOT ${sdk_target_sysroot})
+
+# Specify the cross compiler
+set(toolchain "/opt/poky/3.1.7/sysroots/x86_64-pokysdk-linux/usr/bin/aarch64-poky-linux/")
+set(toolchain_prefix "aarch64-poky-linux-")
+set(triple aarch64-poky-linux)
+set(CMAKE_C_COMPILER_TARGET ${triple})
+set(CMAKE_C_COMPILER ${toolchain}${toolchain_prefix}clang)
```
Comment (HidenoriMatsubayashi) [inline cmake/cross-toolchain.cmake:15]: ${toolchain}${toolchain_prefix} -> ${toolchain}/${toolchain_prefix}

File: cmake/cross-toolchain.cmake [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_SYSTEM_PROCESSOR aarch64)
+
+# Specify the sysroot
+set(sdk_target_sysroot "/opt/poky/3.1.7/sysroots/aarch64-poky-linux")
+set(CMAKE_SYSROOT ${sdk_target_sysroot})
+
+# Specify the cross compiler
+set(toolchain "/opt/poky/3.1.7/sysroots/x86_64-pokysdk-linux/usr/bin/aarch64-poky-linux/")
```
Comment (HidenoriMatsubayashi) [inline cmake/cross-toolchain.cmake:11]: set(toolchain "/opt/poky/3.1.7/sysroots/x86_64-pokysdk-linux/usr/bin/aarch64-poky-linux/")
-> set(toolchain "/opt/poky/3.1.7/sysroots/x86_64-pokysdk-linux/usr/bin/aarch64-poky-linux")

File: cmake/cross-toolchain.cmake (hunk 3 above)
Comment (HidenoriMatsubayashi) [reply cmake/cross-toolchain.cmake:15]: By the way, why do think you need `${toolchain}`? The path to cross-compiler is usually already added by developers. I think it is not nessesary. Also, `${toolchain_prefix}` is also the same because `CMAKE_C_COMPILER_TARGET` is added in this file.

File: cmake/cross-toolchain.cmake [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+set(toolchain_prefix "aarch64-poky-linux-")
+set(triple aarch64-poky-linux)
+set(CMAKE_C_COMPILER_TARGET ${triple})
+set(CMAKE_C_COMPILER ${toolchain}${toolchain_prefix}clang)
+set(CMAKE_CXX_COMPILER_TARGET ${triple})
+set(CMAKE_CXX_COMPILER ${toolchain}${toolchain_prefix}clang++)
+
+# Specify the flags
+set(CFLAGS " -O2 -pipe -g -feliminate-unused-debug-types ")
+set(CXXFLAGS " -O2 -pipe -g -feliminate-unused-debug-types ")
```
Comment (HidenoriMatsubayashi) [inline cmake/cross-toolchain.cmake:21]: What is the purpose you added those flags?

File: meta-flutter/README.md [hunk 6]
```diff
@@ -2,17 +2,33 @@
 Recipe file examples for Yocto Project Yocto Project. See also: https://docs.yoctoproject.org/
 
 ## Setup environment
-### meta-clang
-If needed, add meta-clang layer into conf/bblayers.conf:
+### Create a base environment
+Use the Yocto project version `dunfell` :
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:6]: Why do we need to use dunfell version?

File: meta-flutter/README.md [hunk 7]
```diff
@@ -39,5 +55,33 @@ You need to install libsystemd in the same way with the DRM-GBM backend.
 $ bitbake flutter-drm-eglstream-backend
 ```
 
+## SDK toolchain
+If you use the SDK, please follow the steps below.
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:59]: What is the SDK?

File: meta-flutter/README.md (hunk 6 above)
Comment (HidenoriMatsubayashi) [reply meta-flutter/README.md:6]: You need to separate the explain the environment and building yocto.

File: meta-flutter/README.md [hunk 8]
```diff
@@ -39,5 +55,33 @@ You need to install libsystemd in the same way with the DRM-GBM backend.
 $ bitbake flutter-drm-eglstream-backend
 ```
 
+## SDK toolchain
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:58]: Cross-bulding with Yocto SDK. You need to separate the explain how to build Yocto SDK and cross-building flutter-embedded-linux with it.

Comment (HidenoriMatsubayashi) [reply meta-flutter/README.md:58]: In addition, you need to separate the explain how to build Yocto SDK and cross-building flutter-embedded-linux with it.

File: cmake/cross-toolchain.cmake [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+set(CMAKE_C_FLAGS ${CFLAGS} CACHE STRING "" FORCE)
+set(CMAKE_CXX_FLAGS ${CXXFLAGS}  CACHE STRING "" FORCE)
+set(CMAKE_ASM_FLAGS ${CMAKE_C_FLAGS} CACHE STRING "" FORCE)
+set(CMAKE_LDFLAGS_FLAGS ${CMAKE_CXX_FLAGS} CACHE STRING "" FORCE)
```
Comment (HidenoriMatsubayashi) [inline cmake/cross-toolchain.cmake:25]: Why are these necessary?

File: meta-flutter/README.md [hunk 10]
```diff
@@ -39,5 +55,33 @@ You need to install libsystemd in the same way with the DRM-GBM backend.
...
+```
+See also: [SDK building an sdk installer](https://www.yoctoproject.org/docs/2.1/sdk-manual/sdk-manual.html#sdk-building-an-sdk-installer)
+
+Install the SDK :  
+e.g. aarch64
+```Shell
+$ ./tmp/deploy/sdk/poky-glibc-x86_64-core-image-weston-aarch64-qemuarm64-toolchain-3.1.7.sh
+```
+
+This `cross-toolchain.cmake` is a sample of aarch64. Please set the variable in your environment to match the location of your SDK installation :
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:78]: This `cross-toolchain.cmake` -> `[cross-toolchain.cmake](URI)`

Comment (HidenoriMatsubayashi) [reply meta-flutter/README.md:78]: `Please set the variable in your environment to match the location of your SDK installation`

What does it mean?

File: CMakeLists.txt [hunk 11]
```diff
@@ -1,9 +1,11 @@
...
 project("flutter_linuxes" LANGUAGES CXX C)
 
-# Rules of compiler and cpp version.
-set(CMAKE_C_COMPILER clang)
-set(CMAKE_CXX_COMPILER clang++)
+if(NOT DEFINED CMAKE_TOOLCHAIN_FILE)
+    set(CMAKE_C_COMPILER clang)
+    set(CMAKE_CXX_COMPILER clang++)
+endif()
+# cpp version.
```
Comment (HidenoriMatsubayashi) [inline CMakeLists.txt:8]: Please delete this comment. It's very clear. Rather, you need to add the comment you will fail if you try to build with gcc compiler instead of clang.

File: CMakeLists.txt [hunk 12]
```diff
@@ -1,9 +1,11 @@
 cmake_minimum_required(VERSION 3.10)
 project("flutter_linuxes" LANGUAGES CXX C)
 
-# Rules of compiler and cpp version.
-set(CMAKE_C_COMPILER clang)
-set(CMAKE_CXX_COMPILER clang++)
+# You will fail if you try to build with gcc compiler instead of clang.
```
Comment (HidenoriMatsubayashi) [inline CMakeLists.txt:4]: "We only support clang, so the build will be faild if you try to build with gcc compiler instead of it."

File: doc/README.md [hunk 13]
```diff
@@ -129,6 +129,15 @@ $ cmake -DUSER_PROJECT_PATH=examples/flutter-x11-client ..
...
 ```
 
+### 3.2. Cross-build
+You need to create a toolchain file to cross compile using the Yocto SDK for aarch64 on x64 hosts. [toolcahin-template.cmake](../cmake/cross-toolchain-aarch64-template.cmake) is the templete file for aarch64 toolchain. Also, you need to modify <path_to_user_target_sysroot> appropriately for your environment if you want to use the template file.
+```Shell
+$ cmake -DUSER_PROJECT_PATH=<path_to_user_project> -DCMAKE_TOOLCHAIN_FILE=<toolcahin-template-file>
+```
+
+### 3.3. Yocto-build
+If you want to build using yocto, see: [meta-flutter](../meta-flutter/)
```
Comment (HidenoriMatsubayashi) [inline doc/README.md:139]: Yocto

File: meta-flutter/README.md [hunk 14]
```diff
@@ -1,22 +1,72 @@
 # meta-flutter for Yocto Project
-Recipe file examples for Yocto Project Yocto Project. See also: https://docs.yoctoproject.org/
+Recipe file examples for Yocto Project. See also: https://docs.yoctoproject.org/
 
-## Setup environment
-### meta-clang
-If needed, add meta-clang layer into conf/bblayers.conf:
+## Overview
+As a premise, this document uses Yocto Version(dunfel), Yocto Image(core-image-weston) and Machine Selection(qemuarm64) as an example.
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:5]: Please delte that.

File: meta-flutter/README.md [hunk 15]
```diff
@@ -1,22 +1,72 @@
 # meta-flutter for Yocto Project
-Recipe file examples for Yocto Project Yocto Project. See also: https://docs.yoctoproject.org/
+Recipe file examples for Yocto Project. See also: https://docs.yoctoproject.org/
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:2]: Recipe file examples for Yocto Project. In this README, we explain how to build for aarch64 using `core-image-weston` which is one of Yocto Images, and `dunfel` which is one of LTS Yocto versions. See also: https://docs.yoctoproject.org/

File: CMakeLists.txt (hunk 12 above)
Comment (TomohideMorimoto) [reply CMakeLists.txt]: I fixed it.

File: doc/README.md (hunk 13 above)
Comment (TomohideMorimoto) [reply doc/README.md]: I fixed it.

File: meta-flutter/README.md (hunk 15 above)
Comment (TomohideMorimoto) [reply meta-flutter/README.md]: I fixed it.

File: meta-flutter/README.md (hunk 14 above)
Comment (TomohideMorimoto) [reply meta-flutter/README.md]: Removed Overview section.

File: meta-flutter/README.md [hunk 16]
```diff
@@ -1,22 +1,60 @@
 # meta-flutter for Yocto Project
-Recipe file examples for Yocto Project Yocto Project. See also: https://docs.yoctoproject.org/
+Recipe file examples for Yocto Project. In this README, we explain how to build for aarch64 using core-image-weston which is one of Yocto Images, and dunfell which is one of LTS Yocto versions. See also: https://docs.yoctoproject.org/
+
 
 ## Setup environment
-### meta-clang
-If needed, add meta-clang layer into conf/bblayers.conf:
+There are two ways to build using Yocto. One is a build using bitbake and the other is a build using Yocto SDK. Please set up the build environment for its use.
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:6]: Please delete "Please set up the build environment for its use."

File: meta-flutter/README.md [hunk 17]
```diff
@@ -1,22 +1,60 @@
...
 
-### meta-flutter
+Inside the poky directory, there is a script named oe-init-build-env, which should be used to set up the build environment :
 ```Shell
-$ git clone https://github.com/sony/flutter-embedded-linux.git
-$ bitbake-layers add-layer ./flutter-embedded-linux/meta-flutter/
+$ source poky/oe-init-build-env build
+```
+
+You need to select a specific machine to target the build with. Selecting the following in your conf/local.conf :
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:21]: Set your target machine in your conf/local.conf :

File: meta-flutter/README.md [hunk 18]
```diff
@@ -1,22 +1,60 @@
...
+Downloading the Poky and meta-clang, meta-flutter source code :
 ```Shell
+$ git clone git://git.yoctoproject.org/poky.git -b dunfell
 $ git clone https://github.com/kraj/meta-clang -b dunfell
-$ bitbake-layers add-layer meta-clang
+$ git clone https://github.com/sony/flutter-embedded-linux.git
 ```
 
-### meta-flutter
+Inside the poky directory, there is a script named oe-init-build-env, which should be used to set up the build environment :
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:16]: Setup the build environment using 'oe-init-build-env' script in Poky

File: meta-flutter/README.md [hunk 19]
```diff
@@ -1,22 +1,60 @@
...
+$ bitbake-layers add-layer ../meta-clang
+```
+
+Add meta-flutter layer into conf/bblayers.conf :
+```Shell
+$ bitbake-layers add-layer ../flutter-embedded-linux/meta-flutter
+```
+
+### Building Yocto SDK(Only when using cross-building with Yocto SDK)
+Putting the following in your conf/local.conf :
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:37]: Add the following in your conf/local.conf:

File: meta-flutter/README.md [hunk 20]
```diff
@@ -1,22 +1,60 @@
...
+```
+CLANGSDK = "1"
 ```
+See also: [Adding clang in generated SDK toolchain](https://github.com/kraj/meta-clang/blob/master/README.md#adding-clang-in-generated-sdk-toolchain)
 
-## Building Flutter
-### Wayland backend 
+Building a target Yocto SDK :
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:43]: Build Yocto SDK for cross-building

File: meta-flutter/README.md [hunk 21]
```diff
@@ -1,22 +1,60 @@
...
+```Shell
+$ bitbake core-image-weston -c populate_sdk
+```
+See also: [SDK building an sdk installer](https://www.yoctoproject.org/docs/2.1/sdk-manual/sdk-manual.html#sdk-building-an-sdk-installer)
+
+Installing Yocto SDK on host PC :
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:49]: Install Yocto SDK

File: meta-flutter/README.md [hunk 22]
```diff
@@ -1,22 +1,60 @@
...
+```Shell
+$ bitbake-layers add-layer ../meta-clang
+```
+
+Add meta-flutter layer into conf/bblayers.conf :
+```Shell
+$ bitbake-layers add-layer ../flutter-embedded-linux/meta-flutter
+```
+
+### Building Yocto SDK(Only when using cross-building with Yocto SDK)
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:36]: Building Yocto SDK (Only when using cross-building with Yocto SDK)

File: meta-flutter/README.md [hunk 23]
```diff
@@ -39,5 +77,20 @@ You need to install libsystemd in the same way with the DRM-GBM backend.
 $ bitbake flutter-drm-eglstream-backend
 ```
 
+## Cross-building with Yocto SDK
+Installation of the cross-toolchain creates a cross-toolchain environment setup script in the directory that Yocto SDK was installed. Source the cross-toolchain environment setup file :
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md]: Setup the cross-building toolchain environment using a script that you built and installed.

File: meta-flutter/README.md [hunk 24]
```diff
@@ -39,5 +77,20 @@ You need to install libsystemd in the same way with the DRM-GBM backend.
...
+```Shell
+$ source /opt/poky/3.1.7/environment-setup-aarch64-poky-linux
+```
+
+Setting environment variables :
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:86]: Set the following environment vars to cross-build using clang.

File: meta-flutter/README.md [hunk 25]
```diff
@@ -39,5 +77,20 @@ You need to install libsystemd in the same way with the DRM-GBM backend.
...
+```Shell
+$ export CC=${CLANGCC}
+$ export CXX=${CLANGCXX}
+```
+
+Build is the same way as self-build, see: [3.1. Self-build](../doc/README.md)
```
Comment (HidenoriMatsubayashi) [inline meta-flutter/README.md:92]: After doing that, you can build the embedder as normal like self-building on hosts. It means you don't need to be aware of cross-building. See: [3.1. Self-build](../doc/README.md)

**Review by HidenoriMatsubayashi — COMMENTED** [review]
Please change the sentences
